File: Data Structure II/python/Day 2/merge_intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        res = []
        curr = intervals[0]
        for i in range(1, len(intervals)):
            if intervals[i][0] <= curr[1]:
                curr[1] = max(curr[1], intervals[i][1])
            else:
                res.append(curr)
                curr = intervals[i]
        res.append(curr)
        return res

File: Data Structure II/cpp/Day 2/MergeIntervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
       sort(intervals.begin(), intervals.end());
        vector<vector<int>> ans;
        ans.push_back({intervals[0][0], intervals[0][1]});
        for(int i = 1; i < intervals.size(); i++) {
            vector<int>& last = ans.back();
            if(intervals[i][0] <= last[1]) {
                last[1] = max(last[1], intervals[i][1]);
            } else {
                ans.push_back({intervals[i][0], intervals[i][1]});
            }
        }
        return ans;
        
    }
};